common/src/main/java/com/microsoft/identity/common/internal/cache/BrokerOAuth2TokenCache.java

```diff
@@ -0,0 +1,588 @@
+                clientId,
+                localAccountId
+        );
+
+        Logger.verbose(
+                TAG + methodName,
+                "Result found? ["
+                        + (null != accountRecord)
+                        + "]"
+        );
+
+        // If nothing was returned, check the foci cache...
+        if (null == accountRecord) {
+            Logger.verbose(
+                    TAG + methodName,
+                    "Inspecting FOCI cache..."
+            );
+
+            accountRecord = mFociCache.getAccountWithLocalAccountId(
+                    environment,
+                    clientId,
+                    localAccountId
+            );
+        }
+
+        Logger.verbose(
+                TAG + methodName,
+                "Result found? ["
+                        + (null != accountRecord)
+                        + "]"
+        );
+
+        return accountRecord;
+    }
+
+    @SuppressWarnings(UNCHECKED)
+    @Override
+    public List<AccountRecord> getAccounts(@Nullable final String environment,
+                                           @NonNull final String clientId) {
+        final String methodName = ":getAccounts (2 param)";
+
+        final List<AccountRecord> allAccounts = new ArrayList<>();
+
+        allAccounts.addAll(mAppUidCache.getAccounts(environment, clientId));
+        allAccounts.addAll(mFociCache.getAccounts(environment, clientId));
+
+        for (final OAuth2TokenCache optionalTokenCache : mOptionalCaches) {
+            allAccounts.addAll(optionalTokenCache.getAccounts(environment, clientId));
+        }
+
+        Logger.verbose(
+                TAG + methodName,
+                "Found ["
+                        + allAccounts.size()
+                        + "] accounts."
+        );
+
+        return allAccounts;
+    }
+
+    /**
+     * Broker-only API. Fetches AccountRecords from all provided caches - makes NO GUARANTEES
+     * as to whether or not an AT/RT pair exists for these Accounts.
+     *
+     * @return A List of AccountRecords, may be empty but is never null.
+     */
+    public List<AccountRecord> getAccounts() {
+        final String methodName = ":getAccounts";
+
+        final List<AccountRecord> allAccounts = new ArrayList<>();
+
+        allAccounts.addAll(mAppUidCache.getAccountCredentialCache().getAccounts());
+        allAccounts.addAll(mFociCache.getAccountCredentialCache().getAccounts());
+
+        for (final MsalOAuth2TokenCache optionalTokenCache : mOptionalCaches) {
+            allAccounts.addAll(optionalTokenCache.getAccountCredentialCache().getAccounts());
+        }
+
+        Logger.verbose(
+                TAG + methodName,
+                "Found ["
+                        + allAccounts.size()
+                        + "] accounts."
+        );
+
+        return allAccounts;
+    }
+
+    @SuppressWarnings(UNCHECKED)
+    public AccountDeletionRecord removeAccountFromDevice(@NonNull final AccountRecord accountRecord) {
```

Comment:
Curious to know the use case of this API ? Are we planning to allow app to delete another app's account? Or it is broker internal only

Reply:
I've added this to support a device-wide signout: something like 'Remove Account' from the MS Authenticator account viewer screen.

Reply:
TODO: Add javadoc

Reply:
Resolved in a23818c96b29831bcdfc484222dc94911df94ca3